docs/tutorials/PyTorch.py

```diff
@@ -0,0 +1,322 @@
+
+# We can look at the gradient descent from a more mathematical lense by looking at the graph z = t ** 2:
+
+t = np.linspace(-5, 5, 100)
+plt.plot(t, t ** 2)
+
+# We start on the left hill at the point (-4, 16):
+
+t = np.linspace(-5, 5, 100)
+plt.plot(t, t ** 2)
+plt.scatter(-4, 16)
+plt.show()
+
+# If we plot a tangent line at this point, we see which directions we can go:
+
+# +
+t = np.linspace(-5, 5, 100)
+plt.plot(t, t ** 2)
+
+
+plt.scatter(-4, 16)  # point
+tp = -4
+yp = 16
+
+plt.plot(t, tp * 2 * (t - tp) + yp)  # tangent line
+
+plt.xlim(xmin=-5, xmax=5)
+plt.ylim(ymin=0, ymax=25)
+plt.show()
+# -
+
+# We see we need to go down to go toward the minimum. We take a very small step in the direction of the steepest downward slope. When we take steps, we find the x coordinate of our new location by this equation:
+#
+# $x_{new} = x_{current} - \nabla y(x_{current}) * (step \: size)$
+#
+# where:
+# - $x_{current}$ is our current x value
+#
+# - $\nabla y(x_{current})$ is the gradient at our current point
+#
+# - $(step \: size)$ is a value we choose that tells us scales our steps
+#
+# This makes sense because we start at some x value and we want to get to a new x value that is closer to the minimum. In the above plot, the gradient is negative. If we were to add the gradient to our current x value that would bring us further away from the minimum to a more negative x value. This is because the gradient points in the direction of the steepest ascent while we are looking to go in the direction of the steepest descent. This is why there is a negative in the equation.
+#
+#
+# We will choose `step size = 0.1`:
+
+# +
+t = np.linspace(-5, 5, 100)
+plt.plot(t, t ** 2)
+
+
+plt.scatter(-4, 16)  # point
+
+plt.scatter(-3.2, 10.24)  # second point
+plt.arrow(-4, 16, 0.8, -5.76, width=0.07, color="r")
+plt.text(-4.5, 12, "step 1")
+
+plt.xlim(xmin=-5, xmax=5)
+plt.ylim(ymin=0, ymax=25)
+plt.show()
+# -
+
+# The gradient at our new point is still not close to zero, meaning we haven't reached the minimum. So we will take another step:
+
+# +
+t = np.linspace(-5, 5, 100)
+plt.plot(t, t ** 2)
+
+
+plt.scatter(-4, 16)  # point
+
+plt.scatter(-3.2, 10.24)  # second point
+plt.arrow(-4, 16, 0.8, -5.76, width=0.07, color="r")
+plt.text(-4.5, 12, "step 1")
+
+plt.scatter(-2.56, 6.5536)  # third point
+plt.arrow(-3.2, 10.24, 0.64, -3.6864, width=0.07, color="r")
+plt.text(-4, 7, "step 2")
+
+
+plt.xlim(xmin=-5, xmax=5)
+plt.ylim(ymin=0, ymax=25)
+plt.show()
+# -
+
+# We would continue this process of checking if the gradient is nearly zero, and taking a step in the direction of steepest descent until we reach the bottom of the valley. We'll say we've reached the bottom of the valley when the absolute value of the gradient is <0.1:
+
+# +
+t = np.linspace(-5, 5, 100)
+plt.plot(t, t ** 2)
+
+x_vals = np.array(
```

Comment:
How are the values computed in these arrays? Would it be possible to compute them directly, rather than hard-coding them?

Reply:
Echoing this suggestion because I think the hard-coded values 1) make the process more opaque and 2) quickly become a bit confusing/overwhelming to look at, especially in a tutorial where someone is presumably trying to understand the basics.